Controllers: Add Mapping for Reloop Ready

Swiftb0y opened the pull request:
Initial commit of a mapping for the Reloop Ready controller.
The Padmodes are not implemented and just default to the hotcues
right now. The mapping currently depends on 
 * #4474
 * #4495
 * #4512

Swiftb0y committed:
Controllers: Add Mapping for Reloop Ready

Co-authored-by: majekw <majek@w7i.pl>

majekw commented on res/controllers/Reloop Ready.midi.xml:
```diff
@@ -0,0 +1,1983 @@
+
+            <control>
+                <group></group>
+                <key>ReloopReady.components.leftDeck.fxUnit.level.input</key>
```
```suggestion
                <key>ReloopReady.components.rightDeck.fxUnit.level.input</key>
```
Left deck fx is already in line 281 and there is no rightDeck.fxUnit.level.input anywhere, so knob is not working

majekw commented on res/controllers/Reloop Ready.midi.xml:
```diff
@@ -0,0 +1,1983 @@
+            <control>
+                <group></group>
+                <key>ReloopReady.components.leftDeck.fxUnit.level.input</key>
+                <status>0xB8</status>
```
```suggestion
                <status>0xB9</status>
```
Right value for right fx knob is 0xB9

majekw reviewed (changes requested):
FX unit 2 level knob doesn't work. With suggested changes it works (tested).

Swiftb0y commented:
Thank you very much for your review. Currently there is no-one really reviewing controller mappings apart from myself and I obviously can't review my own mapping, so your effort to review this brings us closer to merging. 
Sorry for the small issues, I'll take care of them ASAP.

majekw commented:
@Swiftb0y, could You add new commits instead of force pushing? I don't know what changed since I tested it last time (no meaningful commit messages or just comments here)  and I don't know if any new testing is required. I own Reloop Ready, but I seldom use it. I know that some modes of performance pads are not implemented yet, so I would be happy to check this as soon as they become available.
At the end You could just use merge with squash if you don't want many small commits merged, or even force push everything

Do you have some checklist to see what should work, what is work in progress and what is still to do?

And thank You for still working on it :+1:

Swiftb0y commented:
@majekw Since this is just a draft PR I wanted to keep the commit history clean. Since this PR contained preliminary commits which were merged in other PRs, I just rebased those out. So now it only contains the two hardware-specific files. The next push will be a couple of fixes and a new feature. I'll explain in the respective commit messages. I'll ping you again once this is ready for more testing.

Swiftb0y commented:
> I know that some modes of performance pads are not implemented yet, so I would be happy to check this as soon as they become available.

I don't think I'll implement the pitch play mode. And for the scratch bank mode, I wanted to repurpose that to beat jumps.

Swiftb0y commented:
@majekw pushed new commit. you can take a quick changelog from the commit description. I developed most things in parallel so splitting into separate commits wouldn't have made sense since I'll squash this anyways.

Swiftb0y commented:
@majekw Were you able to do some testing?

Abdull commented:
@Swiftb0y , thank you for *Reloop Ready* mapping!

I am currently testing your latest mapping version, i.e. 
* https://github.com/Swiftb0y/mixxx/blob/c2c475d4e89ae4e6637bce71fb22a9aa642a715e/res/controllers/Reloop%20Ready.midi.xml
* https://github.com/Swiftb0y/mixxx/blob/c2c475d4e89ae4e6637bce71fb22a9aa642a715e/res/controllers/Reloop-Ready-scripts.js
, against mixxx `2.4-beta-394-gd1c0de51b0`.

I noticed the tempo sliders are behaving inverted (i.e., when physically moving the tempo sliders up towards "minus", actually physically/logically the sliders move down towards "plus").

Setting `invert: true` in the `this.rate` `Pot` component fixes this:

```
    this.rate = new components.Pot({
        inKey: "rate",
        invert: true, // <--- the fix #######################################
        max: 1023,
        // see explanation of ReloopReady.tempoFaderMiddles for info
        inValueScale: function(value) {
            var middle = (this.max + 1) * ReloopReady.tempoFaderMiddles[index];
            return script.absoluteNonLinInverse(value, 0, middle, this.max, 0, 1);
        }
    });
```

Also, I'm trying to also solve an issue with scratch behavior (which is not limited to the Reloop Ready mapping, see https://mixxx.zulipchat.com/#narrow/stream/109122-general/topic/spinning.20platter.20in.20vinyl.20scratch.20mode.20is.20jerky.20.2F.20unsmooth ).

Once you have some spare time, maybe we can put our heads together for solving it.

It be awesome if the Reloop Ready mapping could make it into Mixxx.

Swiftb0y commented:
Hey there @Abdull. I'm afraid I can't do much about the scratch behavior issue (its universal to mixxx and related to the way we handle scratching internally which I'm not familiar nor qualified enough to look into it). I can however help you with the tempo sliders. What are your tempo slider settings (found in the deck preferences)? 
![image](https://github.com/mixxxdj/mixxx/assets/12380386/785093d0-1289-4cfe-8495-c560b641a29d)

Abdull commented:
> Hey there @Abdull. I'm afraid I can't do much about the scratch behavior issue (its universal to mixxx and related to the way we handle scratching internally which I'm not familiar nor qualified enough to look into it). I can however help you with the tempo sliders. What are your tempo slider settings (found in the deck preferences)?

With `invert: true`, everything becomes fine with the tempo sliders.

Regarding scratching, I guess a timeout until no movement is detected would be a viable solution (for all controllers); that is the solution Serato DJ has taken to my observation.

I would have worked on such solution in the past days if only my Reloop Ready hasn't started to logically spin the left jog wheel counterclockwise, no matter if it's spun physically clockwise or counterclockwise.

It's really a cheaply manufactured controller I cannot recommend, including the fact that Reloop is unwilling to provide MIDI mappings when asked.

Swiftb0y commented:
> With invert: true, everything becomes fine with the tempo sliders.

Yeah, I see, but that's weird because that would imply that the messages sent by your controller differ from mine. Can you send me a screenshot for your tempo sliders (in whatever mixxx skin you're using) and from the deck preferences?

> It's really a cheaply manufactured controller I cannot recommend, including the fact that Reloop is unwilling to provide MIDI mappings when asked.

I agree. I already built in some logic to account for units such as mine that have the notch of the tempo fader offset, so it locks the fader at +0.5%bpm speed...

Swiftb0y commented:
put in some love. tested the pitch-play mode and fixed the tempo slider as per @Abdull's suggestion. This is ready to go, I'd appreciate a review.

Swiftb0y commented:
Hi @Abdull, I have made some bigger changes to this mapping. Now based on 2.4. Can you do me a favor and test the current state of the mapping? Thank you.

JoergAtGithub commented on res/controllers/Reloop-Ready-scripts.js:
```diff
@@ -0,0 +1,1141 @@
+    };
+
+
+    for (let i = 0; i < this.pads.length; i++) {
+        this.pads[i] = new components.Button({
+            midi: [0x94 + index, 0x14 + i],
+            group: `[Channel${index + 1}]`,
```
What is i and what is index? Better use self descritive names like padIdx and padModeIdx.

JoergAtGithub commented on res/controllers/Reloop-Ready-scripts.js:
```diff
@@ -0,0 +1,1141 @@
+        }
+    });
+
+    // TODO this seems to have soft-takeover enabled sometimes?!
```
Is this still a ToDo?

Swiftb0y replied:
not sure, I'll investigate

Swiftb0y replied:
okay, it definitely is. I'm not sure why, whether its a mixxx, or a hardware issue and its not deterministically reproducible for me either.

Swiftb0y replied:
done. see comment

JoergAtGithub commented on res/controllers/Reloop-Ready-scripts.js:
```diff
@@ -0,0 +1,1141 @@
+
+    this.pfl = new components.Button({
+        midi: [0x90 + index, 0x1B],
+        shiftOffset: -0xD,
```
Is the negative offset correct here?
Is this really an offset, or an independent absolut value?

Swiftb0y replied:
> Is the negative offset correct here?

Yes, I was confused as well, but thats what it is. 

>  Is this really an offset, or an independent absolut value?

Well, its absolute, but componentJS currently just has the shift offset API, so that's how it is expressed. I didn't want to reimplement the send function just for that.

JoergAtGithub commented on res/controllers/Reloop-Ready-scripts.js:
```diff
@@ -0,0 +1,1141 @@
+ * It uses the lower 6-bit for RGB colors (each channel having the unbelievable
+ * resolution of 2-bit), while the 7th-bit is used to switch between two
+ * brightness/intensity modes (with the 7th-bit set, meaning high brightness).
+ *
```
Good explained!

JoergAtGithub reviewed (commented):
I added some comments - Hope They Help

Swiftb0y commented on res/controllers/Reloop Ready.midi.xml:
```diff
@@ -0,0 +1,2145 @@
+        <name>Reloop Ready</name>
+        <author>Swiftb0y</author>
+        <description>Almost feature-complete mapping of the Reloop Ready hardware. Development kindly sponsored by Samgarr</description>
+        <wiki></wiki>
```
yup, added once merged.

Swiftb0y commented on res/controllers/Reloop-Ready-scripts.js:
```diff
@@ -0,0 +1,1141 @@
+ * @param {(value: number) => void} func callback that is called on ButtonDown
+ * @returns {(channel: number, control: number, value: number, status: number, group: string) => void} XML input handler
+ */
+ReloopReady.makeButtonDownInputHandler = function(func) {
+    return function(channel, control, value, status, _group) {
+        const isPress = this.isPress(channel, control, value, status);
+        this.output(isPress);
+        if (!isPress) {
+            return;
+        }
+        func.call(this, value);
+    };
+};
```
I wouldn't know how to make it simper to be honest. Each step is necessary. How would you simplify it?

Swiftb0y commented on res/controllers/Reloop-Ready-scripts.js:
```diff
@@ -0,0 +1,1141 @@
+// eslint-ignore-next-line no-var
+var ReloopReady = {};
+
+/// The controller only offers Low, High and Gain Knobs,
+/// settings this to true will remap them to control Low, Mid, High instead.
+ReloopReady.threeBandEQ = false;
```
Sure... I don't have much experience with ES6 classes, but I can change it.

JoergAtGithub replied:
Editors like VS-Code can do this conversion automatically

Swiftb0y replied:
how? I haven't seen this feature yet.

JoergAtGithub replied:
It's one of the contexts sensitive functionalities of the light bubble menu:
![grafik](https://github.com/mixxxdj/mixxx/assets/64457745/5ad003fe-6da2-43be-9673-6cc52ac32a5c)

Swiftb0y replied:
Interesting, my version (VSCodium) doesn't offer that. 
![image](https://github.com/mixxxdj/mixxx/assets/12380386/01984478-3599-45a3-93b5-048c896cb488)

Swiftb0y replied:
fyi I don't think it really makes sense to convert this into an ES6 class. At least not the ReloopReady object. The lack of static properties in QJSEngine results in really badly organized code.

Swiftb0y commented on res/controllers/Reloop-Ready-scripts.js:
```diff
@@ -0,0 +1,1141 @@
+                        // non standard behavior
+                        script.bpm.tapButton(index);
+                    } else {
+                        // TODO: For 2.4 using plain beatloop_activate
+                        // is sufficient, see #4328
```
whoops, missed that. On it

Abdull commented:
> Hi @Abdull, I have made some bigger changes to this mapping. Now based on 2.4. Can you do me a favor and test the current state of the mapping? Thank you.

Very nice! Thank you very much.

I just tested `res/controllers/Reloop Ready.midi.xml` and `res/controllers/Reloop-Ready-scripts.js` from https://github.com/mixxxdj/mixxx/pull/4513/commits/d97a241a4ed97802053e0960362bbf0349f14e10 against the Mixxx 2.4.0 release (self-compiled, commit `ca8720f8f365853bb80e23714ed89e41b54d536a`, on Debian bookworm, `uname -a: Linux frami 6.5.0-0.deb12.1-rt-amd64 #1 SMP PREEMPT_RT Debian 6.5.3-1~bpo12+1 (2023-10-08) x86_64 GNU/Linux`).

The tempo faders react logically correctly to physical movement (physical down is logical down, and up is up).

The controller pads shine colorfully according to the hotcue buttons, and CUE blinkenlights nicely.

**The only issue I noticed (with the previous Reloop Ready mapping Aufschlag already) is that when I load a new track on a deck, that deck's physical *VINYL* button LED turns off, even though logically Mixxx correctly has the vinyl mode activated.** This can be workedaround by pressing the deck's VINYL button twice so it shines again.
I can't tell auf die Schnelle if this is a bug with the mapping or a bug with Mixxx.

But besides that, the mapping looks great to me. Thanks @Swiftb0y !

Swiftb0y commented:
> The only issue I noticed (with the previous Reloop Ready mapping Aufschlag already) is that when I load a new track on a deck, that deck's physical VINYL button LED turns off, even though logically Mixxx correctly has the vinyl mode activated. This can be workedaround by pressing the deck's VINYL button twice so it shines again.

I haven't noticed that, but I'll look into it. The vinyl (or more commonly called scratch) button is implemented purely within javascript, so I'm a little confused. What do you mean with "vinyl mode activated [within mixxx]"?

Abdull commented:
> > The only issue I noticed (with the previous Reloop Ready mapping Aufschlag already) is that when I load a new track on a deck, that deck's physical VINYL button LED turns off, even though logically Mixxx correctly has the vinyl mode activated. This can be workedaround by pressing the deck's VINYL button twice so it shines again.
> 
> I haven't noticed that, but I'll look into it. The vinyl (or more commonly called scratch) button is implemented purely within javascript, so I'm a little confused. What do you mean with "vinyl mode activated [within mixxx]"?

Ah okay, TIL vinyl/scratch mode is not modeled within Mixxx, but purely within the controller mappings (so that's probably also why there are no vinyl/scratch mode indicators/buttons within the Mixxx GUI interfaces).

Here is a recording of the vinyl/scratch mode buttons I observe on my Reloop Ready controller:

[![video](https://github.com/mixxxdj/mixxx/assets/529862/cda5b303-afc4-415e-8ecd-591aeaea498f)](https://github.com/mixxxdj/mixxx/assets/529862/cda5b303-afc4-415e-8ecd-591aeaea498f)

As a prerequisite for this glitch to happen, the "SHIFT" button on the Reloop Ready controller has to have been pressed at least once during the Mixxx runtime. From then on, when loading a track to a deck (via pressing "LEFT LOAD" button for left deck, or "RIGHT LOAD" button for right deck), the targeted deck's "VINYL" button light will turn off, even when it is supposed to stay on. 

Maybe it isn't an issue with the Mixxx Reloop Ready controller mapping, but a bug with the Reloop Ready controller firmware &mdash; as per the Windows-based *Reloop Control Panel*, my controller has revision (firmware?) `v1.00`.

![reloop-control-panel-reloop-ready](https://github.com/mixxxdj/mixxx/assets/529862/efdb855a-ea13-4f50-b449-cfb3d4cad5df)

Sending my controller the SysEx MIDI message `F0 00 20 7F 19 F7` (as per Reloop's MIDI mapping PDF `Firmware_map_READY_public_20210311.pdf`, this is the SysEx request for "firmware version") returns `f0 00 20 7f 19 00 01 00 00 f7` &mdash; so this look like a payload of `00 01 00 00` / `v1.00` as well (used WebMIDI tool: https://arachsys.github.io/webmidi/sysex ).

![reloop-ready-sysex-webmidi-arachsys-firmware-version](https://github.com/mixxxdj/mixxx/assets/529862/bcba6ae2-4ab0-46a6-86ad-21d9b83e6ce0)

Swiftb0y commented:
> Here is a recording of the vinyl/scratch mode buttons I observe on my Reloop Ready controller:

Thank you for the in depth video and analysis. The misprint on the box is funny, I've never noticed that. I don't think that is purely graphical however and no real units with that issue exist (hopefully :eyes:). 

> Maybe it isn't an issue with the Mixxx Reloop Ready controller mapping, but a bug with the Reloop Ready controller firmware — as per the Windows-based Reloop Control Panel, my controller has revision (firmware?) v1.00.

Pretty sure its not. FWIW I have the same firmware version. also the newest version of the mapping also requests that from the controller, so you can look it up by inspecting the controller logs (without having to install software only available on windows ;).

I'm fairly certain I know what the issue is and it's pretty straightforward to solve. I'll push a fix soon, I'm just still finishing up the remaining tasks from the previous TODO.

Swiftb0y committed:
various improvements from review and feedback.

ES6 classes,
fix pitch-play with loopcues
removed unused vars
deduplicate loop-related padmodes
query firmware version
softakeover on volume faders
remove pre-2.4 workarounds
fix small looprange

Abdull commented:
FYI (for your inspiration): @andymann has a repository with Mixxx MIDI mappings for the *Reloop Buddy*, which is very similar to the *Reloop Ready*:

https://github.com/Andymann/mixxx-controllers

Swiftb0y commented:
Good to know, I'd love to see that upstreamed. Maybe I can adapt the Ready mapping for the Buddy.

Abdull commented on res/controllers/Reloop-Ready-scripts.js:
```diff
@@ -0,0 +1,1141 @@
+        // midiWheel: [0xB0 + index, 0x06]
+        alpha: 1/8,
+        deck: channel,
+        wheelResolution: 300,
```
I noticed that with these jogwheel settings, it is impossible to set precise cue marks / beatgrids, as already the minimal quantum of jogwheel movement will make the track move by almost 0.01 seconds (maximally zoom into the waveform to see what I mean).

This is true for both vinyl/scratch mode and non-scratch/nudge mode.

The setting of `wheelResolution: 300` and the implied `rpm: 33 + 1/3` do correctly make the spinner GUI rotate one (1) time with one (1) physical jogwheel rotation.

(Each jogwheel has 150 teeth and 150 holes, and two optical sensors, so I guess `wheelResolution: 300` is correct)

But maybe there is another setting or variable that allows for controlling the track movement with the jogwheel in scratch mode and/or non-scratch mode, while still preserving the physically correct spinner GUI movement?

Swiftb0y replied:
While this is valid feedback, I'm unsure how good of a solution a workaround this (essentially hardware) problem is. Though I could imagine slowing down the speed when shift is pressed. Wdyt?

Abdull replied:
I noticed that holding SHIFT while rotating the platter allows for fine track movement. So this is review comment is resolved.

Abdull commented on res/controllers/Reloop-Ready-scripts.js:
```diff
@@ -0,0 +1,1141 @@
+
+ReloopReady.backlightButtons = true;
+
+/// scratch parameter, change to your liking
+ReloopReady.scratchParams = {
+    alpha: 1/8,
+    beta: (1/8)/32,
+};
```
I think `ReloopReady.scratchParams = { ... };` is dead code and can be removed without impact.

Swiftb0y replied:
Yeah, I noticed that when converting the mapping to ES6 classes. Unfortunately that work has stalled and I don't know when I'll be able to finish it. I'll try my best to do so asap. That's also the reason why I haven't pushed any fixes to your previous suggestions yet.

Swiftb0y commented on res/controllers/Reloop Ready.midi.xml:
```diff
@@ -0,0 +1,2143 @@
+        <scriptfiles>
+            <file filename="lodash.mixxx.js"/>
+            <file filename="midi-components-0.0.js"/>
+            <file filename="common-controller-scripts.js"/>
```
Yeah, should work regardless and I included it to make explicit that the mapping depends on it. But the double-inclusion is not great. I think a better fix would rather be to detect this in mixxx though and not double-include then.

Abdull commented on res/controllers/Reloop-Ready-scripts.js:
```diff
@@ -0,0 +1,1141 @@
+                    }
+                }).call(this);
+                engine.setValue(this.group, "pitch_adjust", pitchAdjust);
+                engine.setValue(this.group, `hotcue_${cuepoint}_activate`, value);
```
Great that you got the Pitch Play mode incorporated!

After understanding its usage (pitch shift range can be changed by pressing the *PARAMETER 1* buttons; targeted hotcue can be selected while holding *SHIFT*), I noticed that the Pitch Play behavior may be sort of broken when using it with ***loop** hotcues (*Saved Loops*) because of using [`hotcue_${cuepoint}_activate`](https://manual.mixxx.org/2.4/en/chapters/appendix/mixxx_controls#control-[ChannelN]-hotcue_X_activate) here:

**With a *loop hotcue*, selecting a shift will *toggle* the activation state of the saved loop. This means that for an active hotcue loop, the loop will be deactivated, making the track exit the loop. This is probably not the user's expected behavior and makes the usability complicated.**

This can be worked around e.g. by quickly pressing the upcoming next pitch pad twice (double-toggling), but this is quite complicated for well-timed loop pitching.

From the available `hotcue_X_foo` commands on https://manual.mixxx.org/2.4/en/chapters/appendix/mixxx_controls, I didn't find a suitable alternative to `hotcue_${cuepoint}_activate` that would work well with both normal hotcues and loop hotcues.

But if it is possible to first check the kind of the hotcue, then for a hotcue of type *loop hotcue* to use `engine.setValue(this.group, ``hotcue_${cuepoint}_gotoandloop``, value);` would maybe be a viable solution. The loop can still be deactivated on the Reloop Ready quite userfriendly by pressing the conventional hardware *LOOP* rotary encoder.

Swiftb0y replied:
Yikes, that's some unfortunate feature-interaction. wdyt @Holzhaus?

Abdull replied:
A video is worth a thousand words:

[![]](https://github.com/mixxxdj/mixxx/assets/529862/267bfffa-eb75-4a84-bf3f-94a69fa51b49)

Once "Pitch Play mode" is activated, notice how pressing the pad for the selected *Saved Loop hotcue 5" (white color) with the sole intention of changing the pitch also *toggles* the loop activation state (waveform loop background color toggles from whiteish (deactivated) to whiteish-greenish (activated)).
 
`hotcue_${cuepoint}_activate` feels natural per se in *hot cue mode* for the new concept of *Saved Loops* (i.e., `hotcue_${cuepoint}_activate` toggling the loop activation state for a running track).

It's just that `hotcue_${cuepoint}_activate`'s behavior doesn't fit well for *Pitch Play Mode*.

Abdull replied:
prima: with 61ca002628f244983bad4c7b7ef330a1adface9f, I consider the Pitch Play mode issue to be resolved (I think my account doesn't have the authorization to mark *this*/my review as resolved).

Swiftb0y commented:
new version. see commit message for details. please resolve any review comments that you deem addressed. Thank you.

Swiftb0y committed:
chore: fix eslint

Abdull commented on res/controllers/Reloop-Ready-scripts.js:
```diff
@@ -475,485 +478,467 @@ ReloopReady.AutoLoopPadMode = function(index) {
-        })
-    );
+        }));
+    }
```
I noticed that in *Sampler mode*, when a sample (i.e., track) is selected by pressing a pad, irrespective whether this pad is on the left deck or on the right deck, the sample will get assigned to the same numerical pad on *both* decks.

I don't know if this is the UX behavior as intended by [`components.SamplerButton`](https://github.com/mixxxdj/mixxx/wiki/Components-JS#samplerbutton).

But maybe for the Reloop Ready with its 2 × 8 pads = 16 pads, it would make sense to map the sampler buttons 1 through 8 to the *left* deck pads, and map the sampler buttons 9 through 16 to the *right* deck pads (the *Deere* and *LightNight* skins allow to view 16 samplers through their *skin settings*).

Swiftb0y replied:
I think its a matter of which deck the track is currently playing on. However, I do want to map the parameter buttons to cycle through the different available samplers. would that work for you?

Abdull commented on res/controllers/Reloop-Ready-scripts.js:
```diff
@@ -115,239 +251,121 @@ components.Encoder.prototype.inValueScale = function(value) {
-            unshift: function() {
-                this.inKey = "MoveVertical";
+                this.inKey = "eject";
+                this.outKey = this.inKey;
```
`this.load.shift` is supposed to [`eject`](https://manual.mixxx.org/2.3/en/chapters/appendix/mixxx_controls.html#control-[ChannelN]-eject) a track on the specified deck/channel `deckIdx`, right?

**With the current mapping version, ejecting tracks doesn't work for me, maybe because the related MIDI messages are not set up yet in `Reloop Ready.midi.xml`.**

IMHO the following XML needs to be added to `Reloop Ready.midi.xml`:

```
<!-- ... -->
            <!-- holding shift and pressing left channel's LOAD button to eject track loaded on left channel -->
            <control>
                <group></group>
                <key>ReloopReadyInstance.components.leftChannel.load.input</key>
                <status>0x9E</status>
                <midino>0x1B</midino>
                <options>
                    <script-binding/>
                </options>
            </control>
            <!-- holding shift and UNpressing left channel's LOAD button -->
            <control>
                <group></group>
                <key>ReloopReadyInstance.components.leftChannel.load.input</key>
                <status>0x8E</status>
                <midino>0x1B</midino>
                <options>
                    <script-binding/>
                </options>
            </control>

<!-- ... -->

            <!-- holding shift and pressing right channel's LOAD button to eject track loaded on right channel -->
            <control>
                <group></group>
                <key>ReloopReadyInstance.components.rightChannel.load.input</key>
                <status>0x9E</status>
                <midino>0x10</midino>
                <options>
                    <script-binding/>
                </options>
            </control>
            <!-- holding shift and UNpressing right channel's LOAD button -->
            <control>
                <group></group>
                <key>ReloopReadyInstance.components.rightChannel.load.input</key>
                <status>0x8E</status>
                <midino>0x10</midino>
                <options>
                    <script-binding/>
                </options>
            </control>
<!-- ... -->
```

I haven't understood Mixxx's automagic with respect to [`shiftOffset`](https://github.com/mixxxdj/mixxx/wiki/Components-JS#optional-properties), [`unshift`, and `shift`](https://github.com/mixxxdj/mixxx/wiki/Components-JS#shift-layers). I noticed that the offsets are different for the left-deck-shift-load and the right-deck-shift-load:
* "LEFTLOAD" (`0x02`) versus "SHIFT+LEFTLOAD" (`0x1B`) -> offset is **`0x19`**.
* "RIGHTLOAD" (`0x03`) versus "SHIFT-RIGHTLOAD" (`0x10`) -> offset is **`0x0D`** (as currently used in the code for `shiftOffset`.

I noticed I can comment out `shiftOffset: 0x0D` completely, and the ejecting function still works (as long as the XML part is set up as mentioned before).

Swiftb0y replied:
> I haven't understood Mixxx's automagic with respect to [shiftOffset](https://github.com/mixxxdj/mixxx/wiki/Components-JS#optional-properties), [unshift, and shift](https://github.com/mixxxdj/mixxx/wiki/Components-JS#shift-layers). I noticed that the offsets are different for the left-deck-shift-load and the right-deck-shift-load:

Good catch, I just assumed it was the same. So basically `shift` is called by the enclosing `ComponentContainer` and it reconfigures the component for its shifted behavior, `unshift` is then called on shift-button release (as well as on component init to avoid some code duplication). `shiftOffset` is currently only used for LED-output (as no input binding is done via JS (though #12781 would allow us to change that in the future). 

> this.load.shift is supposed to [eject](https://manual.mixxx.org/2.3/en/chapters/appendix/mixxx_controls.html#control-%5BChannelN%5D-eject) a track on the specified deck/channel deckIdx, right?
> 
> With the current mapping version, ejecting tracks doesn't work for me, maybe because the related MIDI messages are not set up yet in Reloop Ready.midi.xml.

Yup, good catch.

Swiftb0y replied:
Fixed.

JoergAtGithub commented on res/controllers/Reloop-Ready-scripts.js:
```diff
@@ -0,0 +1,1146 @@
+
+// eslint-disable-next-line no-var, no-unused-vars
+var ReloopReadyInstance = {
+    init: function() {
+        ReloopReady.init();
+
+    }
+};
+
+class ReloopReady {
+    static init() {
+        ReloopReadyInstance = new ReloopReady();
+    }
```
This looks rather complicated. Could you check if the code of the Z1 mapping would work here as well:
https://github.com/mixxxdj/mixxx/blob/23e91ac0bf7c86c5b03dc2d7048396be6de2ad5c/res/controllers/Traktor-Kontrol-Z1-scripts.js#L266

Swiftb0y replied:
That's what I would've preffered too, but it didn't work out due to naming issues when I tried. I'll try again soon though.

Swiftb0y replied:
Okay, So the problem is two-fold: I'm doing the `.init` work in the constructor (which I'd like to keep as it makes most sense logically IMO). So I can't construct the instance here. The other problem is that simply assigning `var ReloopReadyInstance = ReloopReady;` won't work before the class has been defined. That would force me to move that `var` decl further down in the file which goes against the pattern established in all other mappings so far. Thus I conclude that the current code is optimal. Wdyt?

Abdull commented on res/controllers/Reloop-Ready-scripts.js:
```diff
@@ -475,485 +478,467 @@ ReloopReady.AutoLoopPadMode = function(index) {
-            on: ReloopReady.padColorPalette.Green.lit,
-            outConnect: false,
-        });
+ReloopReady.LoopRollPadMode = class extends ReloopReady.AbstractLoopPadMode {
```
Regression? The funky *Loop Roll* mode doesn't work since this commit. It worked before.

Swiftb0y replied:
not sure. I'll look into it once I have access to the hardware again.

Swiftb0y replied:
I have the hardware again and I can confirm this. Looking into it.

Swiftb0y committed:
fix: LoopRollPadMode not working

Swiftb0y committed:
refactor: implement parameter buttons for all `AbstractLoopPadMode`s

Swiftb0y committed:
refactor: optimize ColorMapper initialization

Swiftb0y commented:
Fixed. I'd highly appreciate if you could do some testing @Abdull. I've also rebased this PR on 2.5. While it shouldn't require any 2.5 features right now, I will push add the new controller settings soon, at which point you will need a 2.5 installation.

Swiftb0y committed:
wip

Swiftb0y committed:
fix: midiToFullColor algorithm

Swiftb0y committed:
feat: convert to controller settings GUI

Swiftb0y committed:
refactor: remove dependency on lodash

Swiftb0y committed:
polish: ensure that padmode parameter buttons work in all modes

Swiftb0y committed:
fix: shift+loadbutton for eject input and led feedback

Abdull commented:
> Fixed. I'd highly appreciate if you could do some testing @Abdull.

Thanks, I can give it a good round of testing next year, i.e. tomorrow or the day after it ;). I will report.

Swiftb0y committed:
fix: add correct no-unused-vars eslint exception

Swiftb0y committed:
chore: avoid partially unsupported _ in number literals

Swiftb0y committed:
remove outdated TODOs; add manual link; remove duplicate script-file inclusions

Swiftb0y commented:
all green. Already did some testing and I also made sure I didn't miss anything from your previous review comments. Looking forward to your testing @Abdull

Swiftb0y committed:
feat: bind shift+parameter1/2 buttons; add focused hotcue color ctrl

Abdull commented:
I executed the following commands to build _this_ PR:
 
```
ccache --set-config=max_size=20.0G
# check active user-wide ccache configuration with
#   ccache -p

mkdir -p "${HOME}/git"
cd "${HOME}/git"

git clone --recurse-submodules https://github.com/mixxxdj/mixxx

cd "${HOME}/git/mixxx"

# checkout _this_ PR
gh pr checkout 4513

# I use Debian 12 btw
tools/debian_buildenv.sh setup

cmake -DCMAKE_INSTALL_PREFIX=/usr/local -S ~/git/mixxx -B ~/mixxx/git/build

cmake --build ~/git/mixxx/build --parallel $(nproc)

sudo cmake --build ~/git/mixxx/build --target install --parallel $(nproc)
```

Next, I cleaned up/removed manually-added files from "${HOME}/.mixxx/controllers" (they overlay/shadow/win over the same-filenamed versions found under `${HOME}/git/mixxx/res/controllers`, having caused issues for me in the past by Mixxx using a mishmash of different controller file versions not working together; not sure about the winning precedence when also considering `/usr/local/share/mixxx/controllers` ) .

Executing `"${HOME}/git/mixxx/build/mixxx"` starts up the expected version `2.5.0-21-g4df4f3c29e (reloop-ready)`.

Successfully checked that the "Mapping Files" (blue clickable file links) are indeed pointing to the most recent versions from _this_ pull request (i.e., all these "built-in" files are under `${HOME}/git/mixxx/res/controllers`).

Happy to see the additional GUI tweaking options "Mapping settings" and "Tempo Fader neutral correction (See Manual)". I will set Channel Knob Assignment to "Filter, Low, Mid, High". Awesome!

Next up, I will do some mixing and report on that.

Swiftb0y commented:
Great. Thank you for the feedback. There are no C++ changes in this PR so you also could've simply copied the files to your `~/.mixxx/controllers/` directory and used with 2.5 stable.

Abdull commented:
> all green. Already did some testing and I also made sure I didn't miss anything from your previous review comments. Looking forward to your testing @Abdull

Observations so far:
* For deck 1, despite its *VINYL* mode activated and the deck playing, forward-/backward-fast-spinning does not work: deck 1 stays in "speed-up-slow-down" mode (don't know the proper term for non-vinyl mode right now). This issue does not present itself on deck 2.
* @Swiftb0y, if I remember correctly, you also experienced the issue of the Reloop Ready crossfader spuriously software-wise sticking in place despite being physically moved. I still experience this phenomenon. When this happens, a new track on e.g. deck 2 I want to punch in (by moving the crossfader physically e.g. from complete left to center position) stays silent (the crossfader logically stays to the complete left). I think this is a feature of Mixxx and other DJ software to avoid abrupt value jumps for e.g. volume faders, so that the operator first has to physically slide over the currently set logical value in order to "pick up"/attach physical to logical control. But with the Reloop Ready, that's somehow a problem (maybe due to hardware noise). I also remember you did have some source code-controlled option in the Reloop Ready mapping files to deactivate this feature.  Maybe you can expose it as an option to deactivate the "pick up feature" within the GUI configuration for all kinds of range values (?
* Vinyl / scratch mode (as it currently works only on deck 2) still produces a wobbly, unnatural movement when backspinning/forwardspinning _after_ fingers are taken off the platter, manifesting acoustically like a worn-out audio tape. I think this is a bug/issue present in all controller mappings (not only in this Reloop Ready mapping). Curiously, the wobbly/jerky movement does not appear in "non-vinyl"/"vinyl deactivated" mode when the track is paused, albeit the non-vinyl mode produces distinct other sounds (this formant-like "keeping almost stable pitch" sound). In the past, I tried to fix this "wobbly fingers-off vinyl movement" in your Reloop Ready JavaScript coe with the following workaround: despite fingers off the platter, keep state of this platter as if still touched (because for whatever reasons, this does not cause wobbly spinning); then when spinning speed goes below some threshold, restore actual state of no fingers on platter. Obviously, while this did work sort of, it just caused further issues. I wish I better understood Mixxx' model of writing mappings so I could contribute a working solution.
* in the GUI configuration, "Tempo Fader neutral correction (See Manual)": do you mean as _manual_ the comment in `Reloop Ready.midi.xml`? I.e.:

```
            <!--
                On my hardware unit, the tempo faders have such a cheap build-quality
                that the notches in the center are not actually in the middle
                If you find your notches are offset too, do this:
                Load this mapping in mixxx using `developer` mode.
                Then move the tempo fader into the notch at the position where
                the tempo should then be the true tempo (+/- 0). Open the developer tools
                and read the `rate` value of the channel. If its already 0.5 you should be
                good, if its not that, copy the value here and the fader will be adjusted
                for you.
            -->
```

Swiftb0y commented:
Thanks a lot for the thorough report. I will try to address them ASAP.

> For deck 1, despite its VINYL mode activated and the deck playing, forward-/backward-fast-spinning does not work: deck 1 stays in "speed-up-slow-down" mode (don't know the proper term for non-vinyl mode right now). This issue does not present itself on deck 2.

Thats weird, I'll look into it.

> if I remember correctly, you also experienced the issue of the Reloop Ready crossfader spuriously software-wise sticking in place despite being physically moved. 

I experienced it with the volume faders IIRC. The feature is called "soft-takeover" and its easy to disable in this case. I'll push a commit soon.

> Vinyl / scratch mode (as it currently works only on deck 2) still produces a wobbly, unnatural movement when backspinning/forwardspinning after fingers are taken off the platter, manifesting acoustically like a worn-out audio tape. I think this is a bug/issue present in all controller mappings (not only in this Reloop Ready mapping). 

Yeah, its an issue in the scratch code no one has bothered to look into. Thats out-of-scope for now IMO. The way the scratching is done is the mapping is correct, its just not calculated correctly in the actual C++ scratch code. 

> in the GUI configuration, "Tempo Fader neutral correction (See Manual)": do you mean as manual the comment in Reloop Ready.midi.xml?

No I mean manual as in the (broken until the corresponding https://github.com/mixxxdj/manual/pull/478 is merged) "Mixxx Manual" link at the top right (next to "Support").

Swiftb0y committed:
fix: disable softtakeover for crossfader

This avoids an issue where it could get detached when moved quickly.
This can happen often on this controller due to the cheap built
quality of the faders, resulting in low-resolution data that is
larger than the softtakeover treshold.

Swiftb0y commented:
>> For deck 1, despite its VINYL mode activated and the deck playing, forward-/backward-fast-spinning does not work: deck 1 stays in "speed-up-slow-down" mode (don't know the proper term for non-vinyl mode right now). This issue does not present itself on deck 2.
>
> Thats weird, I'll look into it.

I can't confirm this. Without further debugging, my best guess would be a hardware problem on your side. Can you double check whether you have the correct `ReloopReadyInstance.components.leftDeck.jog.inputTouch` mappings in your XML? Also double check whether mixxx (using `--controller-debug`) receives messages when you touch and release messages (specifically `status 0x90 (ch 1, opcode 0x9), ctrl 0x06` messages). Does the issue go away when you toggle the `vinyl` button a couple times?

Swiftb0y committed:
refactor: use util methods for vinyl button

majekw commented:
Deck 1 works as intended in vinyl mode on my unit.

Another thing: there is slight delay between touching jog wheel and Mixxx reaction and stratch effect. I see that touch events from controller are instant, so delay is somewhere in Mixxx. Is it bug or should I get used to this delay?

Swiftb0y commented:
Thanks for the testing Marek.

> Another thing: there is slight delay between touching jog wheel and Mixxx reaction and stratch effect. I see that touch events from controller are instant, so delay is somewhere in Mixxx.

If the delay is only audible (and also happens when scratching on the waveform) a too large audio output buffer size may be at fault. Apart from that, it may be some weird behavior with the underlying mixxx jog code again... I'm not sure. For now I'm afraid you're gonna have to get used to it.

Abdull commented:
> > > For deck 1, despite its VINYL mode activated and the deck playing, forward-/backward-fast-spinning does not work: deck 1 stays in "speed-up-slow-down" mode (don't know the proper term for non-vinyl mode right now). This issue does not present itself on deck 2.
> > 
> > 
> > Thats weird, I'll look into it.
> 
> I can't confirm this. Without further debugging, my best guess would be a hardware problem on your side. Can you double check whether you have the correct `ReloopReadyInstance.components.leftDeck.jog.inputTouch` mappings in your XML? Also double check whether mixxx (using `--controller-debug`) receives messages when you touch and release messages (specifically `status 0x90 (ch 1, opcode 0x9), ctrl 0x06` messages). Does the issue go away when you toggle the `vinyl` button a couple times?

Here is a video of the issue (scratch/vinyl not working on deck 1) I have: https://youtube.com/shorts/81-jZeelp4c

Abdull commented:
> Thanks. Unfortunately the video is too blurry to see the log messages on screen, so I can't tell whats wrong. Can you post a small log where you scratch once with the working wheel and once with the broken one?

It turned out that my Reloop Ready device has/had a hardware fault: the capacitive sensing on the left jog wheel was not physically registered, hence the following MIDI events did not get emitted by the device:

```
            <control>
                <group></group>
                <key>ReloopReadyInstance.components.leftDeck.jog.inputTouch</key>
                <status>0x90</status>
                <midino>0x06</midino>
                <options>
                    <script-binding/>
                </options>
            </control><control>
                <group></group>
                <key>ReloopReadyInstance.components.leftDeck.jog.inputTouch</key>
                <status>0x80</status>
                <midino>0x06</midino>
                <options>
                    <script-binding/>
                </options>
            </control>
```

The issue disappeared after I unscrewed all outside screws and took a peek inside.

So excuse the confusion; yet another hardware issue with the Reloop Ready to be aware of.

Swiftb0y commented:
Great. Thanks for confirming. I'm currently quite swamped on tasks, so if you want to lend me a hand, I want to kindly ask you to help me get this over the finish line. While the mapping is fairly complete now, the [manual page is still very much incomplete](https://github.com/mixxxdj/manual/pull/478). Are you interested in helping with that?

acolombier commented:
I'm marking this PR as draft as the manual is being completed, but let me know when this is ready and we can merge this one!

Swiftb0y commented:
Sure. I'll see if I can revive that.
